workflow.go

```diff
@@ -0,0 +1,257 @@
+				return nil
+			}
+		}
+
+		s.TrialPeriodExpiration = time.Time{}
+		err = workflow.ExecuteActivity(ctx, a.SendTrialExpiredEmail, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("SendTrialExpiredEmail Activity failed.", "Error", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *SubscriptionInfo) billingCycle(ctx workflow.Context) (err error) {
+	logger := workflow.GetLogger(ctx)
+	selector := workflow.NewSelector(ctx)
+	info := workflow.GetInfo(ctx)
+
+	var a *Activities
+	cancelSub := false
+
+	for !cancelSub {
+		// start of billing cycle, charge a subscription
+		err = workflow.ExecuteActivity(ctx, a.ChargeSubscription, s).Get(ctx, nil)
+		if err != nil {
+			logger.Error("ChargeSubscription Activity failed.", "Error", err)
+			return err
+		}
+
+		timerDuration := s.BillingPeriodDuration
+		if !s.BillingPeriodExpiration.IsZero() {
+			timerDuration = s.BillingPeriodExpiration.Sub(workflow.Now(ctx))
+		} else {
+			s.BillingPeriodExpiration = workflow.Now(ctx).Add(timerDuration)
+		}
+		billingTimer := workflow.NewTimer(ctx, timerDuration)
+		var timerError error
+
+		selector.
+			AddFuture(billingTimer, func(f workflow.Future) {
+				// start timer for billing period. Must be async as update/cancel signals may come in.
+				logger.Info("Billing cycle is up! Time to charge...")
+				timerError = f.Get(ctx, nil)
+			}).
+			// Signal Handler for updating billing address or other customer info
+			AddReceive(workflow.GetSignalChannel(ctx, "update"), func(c workflow.ReceiveChannel, _ bool) {
+				var updateInfo UpdateSignal
+				c.Receive(ctx, &updateInfo)
+				logger.Info("Received update signal", "data", updateInfo)
+				s = &updateInfo.Subscription
+			}).
+			// Signal handler for canceling the subscription
+			AddReceive(workflow.GetSignalChannel(ctx, "cancel"), func(c workflow.ReceiveChannel, _ bool) {
+				var cancel UpdateSignal
+				c.Receive(ctx, &cancel)
+				logger.Info("Received cancel signal", "data", cancel)
+
+				cancelSub = cancel.CancelSubscription
+			})
+
+		err = workflow.Await(ctx, func() bool {
+			logger.Info("Current history length", "History Length", info.GetCurrentHistoryLength())
+			return info.GetCurrentHistoryLength() > EventHistoryThreshold || selector.HasPending()
+		})
+		if err != nil {
+			logger.Error("Awaiting history or selector failed.", "Error", err)
+			return err
+		}
+
+		if info.GetCurrentHistoryLength() > EventHistoryThreshold {
+			canError := workflow.NewContinueAsNewError(ctx, SubscriptionWorkflow, *s)
+			return canError
+		}
+
+		selector.Select(ctx)
+
+		if timerError != nil {
+			logger.Error("Billing timer failed.", "Error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *SubscriptionInfo) drainSignals(ctx workflow.Context) (*SubscriptionInfo, error) {
+	// If there are *any* cancel signals,
+	signalChannel := workflow.GetSignalChannel(ctx, "cancel")
+	var cancel UpdateSignal
+	ok := signalChannel.ReceiveAsync(&cancel)
+	if ok && cancel.CancelSubscription {
```

Comment:
Generally to the point of `UpdateSignal` reuse, why can you accept a cancel signal that doesn't have `CancelSubscription` as true? Seems like it shouldn't even be an option.